typegen/src/utils.rs

```diff
@@ -81,7 +81,10 @@ pub fn ensure_unique_type_paths(types: &mut PortableRegistry) {
 /// all type ids mentioned in the TypeDef are either:
 /// - equal
 /// - or different, but map essentially to the same generic type parameter
-fn types_equal_extended_to_params(a: &Type<PortableForm>, b: &Type<PortableForm>) -> bool {
+pub(crate) fn types_equal_extended_to_params(
```

Comment:
So right now this function Does a shallow compare, and to use an  earlier example an issue is eg if we compare:

```rust
struct Foo<T = u64> {
    inner: Vec<u64>
}

struct Foo<T = u32> {
    inner: Vec<u32>
}
```

We would return false because `inner` would be pointing to a different type ID in either case, and require those types to be deduplicated into eg `Foo` and `Foo2`.

We could be smarter and notice that the `T` in `Foo<T>` is used as a param in `Vec` and so we only need to output one type during codegen, but it's more complicated than just recursively calling `types_equal_extended_to_params` because we need to track usages of the generic type, and I'm not sure if that's possible offhand?

Ie we could note that since `T` has type ID 32, then any instances of that ID are actually instances of T being used, but that might not be the case (eg struct `Bar<T = u32> { a: T, b: u32 }`). We could maybe use the `typeName` if that is equal to `T` too or something, in conjunction with the type ID? I'm curious if you have any ideas here @tadeohepperle!

All that said, for now it's perfeclty OK if it is more conservative as long as `ensure_unique_type_paths` then uses this to rename the type paths appropriately. Better to have more duplicated types being spat out in the codegen than to assume types can be merged when in fact they can't!